### PR TITLE
be/interpreter2: an interpreter based on fuir

### PR DESCRIPTION
**Pull request by michaellilltokiwa**
Basically almost everything was just copied over from interpreter. 
So this is not yet base on fuir but already uses the abstract interpreter for execution.
My recommendation to start reviewing this PR is by reading `Executor.java`

### CODE REVIEW

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 1]
```diff
@@ -0,0 +1,1534 @@
...
+                  open_results[0] = _openStreams_.add(fos);
+                  break;
+                case 2:
+                  RandomAccessFile fas = new RandomAccessFile(utf8ByteArrayDataToString(args.get(1)), "rw");
+                  fas.seek(fas.length());
+                  open_results[0] = _openStreams_.add(fas);
+                  break;
+                default:
+                  open_results[1] = -1;
+                  System.err.println("*** Unsupported open flag. Please use: 0 for READ, 1 for WRITE, 2 for APPEND. ***");
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:498]:
```suggestion
                  say_err("*** Unsupported open flag. Please use: 0 for READ, 1 for WRITE, 2 for APPEND. ***");
```

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 2]
```diff
@@ -0,0 +1,1534 @@
...
+    put("fuzion.java.Java_Object.is_null0", (interpreter, innerClazz) -> args ->
+        {
+          Instance thizI = (Instance) args.get(0);
+          Object thiz = JavaInterface.instanceToJavaObject(thizI);
+          return new boolValue(thiz == null);
+        });
+    putUnsafe("fuzion.java.get_static_field0",
+        "fuzion.java.get_field0"      , (interpreter, innerClazz) ->
+        {
+          String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzIntrinsicName(cl);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:719]:
```suggestion
          String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzOriginalName(cl);
```

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 3]
```diff
@@ -0,0 +1,1534 @@
...
+              Object thiz  = statique  ? null :          JavaInterface.instanceToJavaObject(clazzOrThizI);
+              String field = (String) JavaInterface.instanceToJavaObject(fieldI);
+              return JavaInterface.getField(clazz, thiz, field, resultClazz);
+            };
+        });
+    putUnsafe("fuzion.java.call_v0",
+        "fuzion.java.call_s0",
+        "fuzion.java.call_c0", (interpreter, innerClazz) ->
+        {
+          String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzIntrinsicName(cl);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:736]:
```suggestion
          String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzOriginalName(cl);
```

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 4]
```diff
@@ -0,0 +1,1534 @@
...
+                                   /* type  */ ArrayData.elementType(innerClazz._outer));
+        });
+    put("fuzion.sys.internal_array.setel", (interpreter, innerClazz) -> args ->
+        {
+          ((ArrayData)args.get(1)).set(
+                              /* index */ args.get(2).i32Value(),
+                              /* value */ args.get(3),
+                              /* type  */ ArrayData.elementType(innerClazz._outer));
+          return Value.EMPTY_VALUE;
+        });
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:876]: See my comment for `ArrayData.alloc`, these should pass the element clazz to `ArrayData.alloc`.

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 5]
```diff
@@ -0,0 +1,1534 @@
...
+    put("fuzion.sys.env_vars.set0"  , (interpreter, innerClazz) -> args -> new boolValue(false));
+    // unsetting env variable not supported in java
+    put("fuzion.sys.env_vars.unset0", (interpreter, innerClazz) -> args -> new boolValue(false));
+    put("fuzion.sys.misc.unique_id",(interpreter, innerClazz) -> args -> new u64Value(_last_unique_id_.incrementAndGet()));
+    put("fuzion.sys.thread.spawn0", (interpreter, innerClazz) -> args ->
+        {
+          var call = Types.resolved.f_function_call;
+          var oc = innerClazz.argumentFields()[0].resultClazz();
+          var ic = oc.lookup(call);
+          var t = new Thread(() -> interpreter.callOnInstance(ic._idInFUIR, new Instance(ic), args.get(1), new List<>(), false /* NYI could we have pre-cond here? */));
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:897]: `Function.call` does not (and should never) have a pre-condition, I think the comment can go.

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 6]
```diff
@@ -0,0 +1,1534 @@
...
+   *
+   * @return a Callable instance to execute the intrinsic call.
+   */
+  static Callable effect(Interpreter2 interpreter, Clazz innerClazz)
+  {
+    return (args) ->
+      {
+        var m = args.get(0);
+        var cl = innerClazz._outer;
+        String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzIntrinsicName(cl);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:1494]:
```suggestion
        String in = innerClazz.feature().qualifiedName();   // == _fuir.clazzOriginalName(cl);
```

File: src/dev/flang/be/interpreter2/Intrinsics.java [hunk 7]
```diff
@@ -0,0 +1,1534 @@
...
+          case "effect.default": if (FuzionThread.current()._effects.get(cl) == null) { check(m != Value.EMPTY_VALUE); FuzionThread.current()._effects.put(cl, m   ); } break;
+          case "effect.abortable" :
+            {
+              var prev = FuzionThread.current()._effects.get(cl);
+              FuzionThread.current()._effects.put(cl, m);
+              var call = Types.resolved.f_function_call;
+              var oc = innerClazz.actualGenerics()[0]; //innerClazz.argumentFields()[0].resultClazz();
+              var ic = oc.lookup(call);
+              try {
+                var ignore = interpreter.callOnInstance(ic._idInFUIR, new Instance(ic), args.get(1), new List<>(), false /* NYI could we have pre-cond here? */);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Intrinsics.java:1507]: `Function.call` does not have a pre-condition, NYI-comment can go.

File: src/dev/flang/be/interpreter2/JavaInterface.java [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+    i = 1;
+    while (d.charAt(i)!=')')
+      {
+        int e = getEnd(d,i);
+        result[cnt] = str2type(d.substring(i,e));
+        cnt++;
+        i = e;
+      }
+    return result;
+  }
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/JavaInterface.java:195]: These apparently have been moved to `dev.flang.util.JavaInterface`, should not be copied here.

File: src/dev/flang/be/interpreter2/LValue.java [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+   */
+  public LValue(Clazz c, Instance cont, int off)
+  {
+    super(c);
+
+    if (PRECONDITIONS) require
+      (cont != null,
+       off >= 0 || c.isUnitType(),
+       c.isUnitType() || off < Layout.get(cont._clazz).size(),
+       c.isUnitType() || off < cont.refs.length);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/LValue.java:76]: Vertical alignment helps understanding this:
```suggestion
       c.isUnitType() || off >= 0,
       c.isUnitType() || off < Layout.get(cont._clazz).size(),
       c.isUnitType() || off < cont.refs.length);
```
Have you added the conditions or have they been remove in the main branch for some reason?

File: src/dev/flang/be/interpreter2/LValue.java [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+   * toString
+   *
+   * @return
+   */
+  public String toString()
+  {
+    return "lvalue[" + container + "@" + offset + "(" + _clazz + ")]" +
+      (_clazz == Clazzes.i32.getIfCreated() ? " (" + i32Value() + ")" :
+       _clazz == Clazzes.u32.getIfCreated() ? " (" + u32Value() + ")" :
+       _clazz == Clazzes.u8.getIfCreated() ? " (" + u8Value() + ")" :
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/LValue.java:344]: You have added `u8` here. Might also make sense to add `i8`, `i16`, `u16`, `f32`, `f64`.

File: src/dev/flang/be/interpreter2/u16Value.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  /**
+   * Store this value in a field
+   *
+   * @param slot the slot that addresses the field this should be stored in.
+   *
+   * @param size the size of the data to be stored
+   */
+  void storeNonRef(LValue slot, int size)
+  {
+    if (size != 1) System.out.println("Assigning "+this);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/u16Value.java:103]: Old debug code, should be removed.

File: src/dev/flang/be/interpreter2/u32Value.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  /**
+   * Store this value in a field
+   *
+   * @param slot the slot that addresses the field this should be stored in.
+   *
+   * @param size the size of the data to be stored
+   */
+  void storeNonRef(LValue slot, int size)
+  {
+    if (size != 1) System.out.println("Assigning "+this);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/u32Value.java:100]: old debug code, should be removed.

File: src/dev/flang/be/interpreter2/u8Value.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  /**
+   * Store this value in a field
+   *
+   * @param slot the slot that addresses the field this should be stored in.
+   *
+   * @param size the size of the data to be stored
+   */
+  void storeNonRef(LValue slot, int size)
+  {
+    if (size != 1) System.out.println("Assigning "+this);
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/u8Value.java:103]: old debug code, should be removed.

File: src/dev/flang/be/interpreter2/Value.java [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+  public int tag()
+  {
+    throw new Error("value "+ this + " of class " + getClass() + " is not a tag");
+  }
+
+
+  public void dump()
+  {
+    System.out.println(toString());
+  }
```
Comment (fridis) [inline src/dev/flang/be/interpreter2/Value.java:322]: remove, seems to be unused.

**Review by fridis — CHANGES_REQUESTED** [review]
Very nice work. 

Found a number of minor points when comparing with the old interpreter, many of them are already issues in the old interpreter but can be fixed now.

File: src/dev/flang/be/interpreter2/LValue.java (hunk 9 above)
Comment (michaellilltokiwa) [reply src/dev/flang/be/interpreter2/LValue.java:76]: Yes, I added these conditions

File: src/dev/flang/be/interpreter2/ArrayData.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+
+  static AbstractType elementType(Clazz arrayClazz)
+  {
+    // NYI: Properly determine generic argument type of array
+    var arrayType = arrayClazz._type;
+    if (arrayType.compareTo(Types.resolved.t_Const_String) == 0 /* NYI: Hack */)
+      {
+        return Types.resolved.t_i32;
```
[the remark this answers is not in the document]
Comment (michaellilltokiwa) [reply src/dev/flang/be/interpreter2/ArrayData.java:188]: I also do not understand it.

File: src/dev/flang/be/interpreter2/Excecutor.java [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+  {
+    return new Pair<Value,Object>(v, null);
+  }
+
+  @Override
+  public Pair<Value, Object> box(Value v, int vc, int rc)
+  {
+    var rcc = _fuir.clazzForInterpreter(rc);
+    var vcc = _fuir.clazzForInterpreter(vc);
+    return pair(new Boxed(rcc, vcc, v /* .cloneValue(vcc) */));
```
[the remark this answers is not in the document]
Comment (michaellilltokiwa) [reply src/dev/flang/be/interpreter2/Excecutor.java:243]: I don't know.

**Comment by michaellilltokiwa** [comment]
I did a test run on flang_dev/content and idioms. They all seem to work.

**Review by fridis — APPROVED** [review]

**Pull request merged**